### PR TITLE
Add microcosms to the website

### PR DESCRIPTION
**Pull request by openbrian**
Microcosms is a feature to support (local) groups of mappers and the events they participate in.

At this time the PR is just for early feedback and discussion.

All of this code is tested using cucumber.  To test run bundle install; bundle exec rake cucumber.

Please submit bugs and feature requests at this repo https://github.com/openstreetmap/openstreetmap-website.

### CODE REVIEW

**Comment by tomhughes** [comment]
Do you not think it might have been a good idea to write the tests using the existing test framework...

I mean sure if we were starting now we might move to cucumber and it might even be a move we should make in general but I'd really prefer not to have to support multiple test frameworks!

**Comment by tomhughes** [comment]
I'd also point out that we generally advise opening discussions on large scale features before doing lots of work on them so that basic structural issues, naming, etc can be resolved before lots of code is written.

I mean this is definitely something that people want, but whether `Microcosm` is the right name is something we need to discuss but would now involve a large amount of work.

File: app/assets/javascripts/event_attendances.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (tomhughes) [inline app/assets/javascripts/event_attendances.coffee]: This, and a whole load of other empty files that have been added automatically by a generator, should be removed as they're not actually being used. Especially since we don't use coffescript!

File: app/helpers/event_attendances_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module EventAttendancesHelper
+end
```
Comment (tomhughes) [inline app/helpers/event_attendances_helper.rb]: Empty helpers can go as well.

File: app/models/application_record.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# Base ApplicationRecord Class
+class ApplicationRecord < ActiveRecord::Base
+  self.abstract_class = true
+end
```
Comment (tomhughes) [inline app/models/application_record.rb]: We don't currently use `ApplicationRecord` so please don't introduce it like this - just make your models inherit in the same way as existing ones.

If you want to introduce `ApplicationRecord` then we can discuss that but it should be done as a separate task and introduced globally for consistency.

File: app/views/events/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    border: 1px solid #888;
+    border-radius: 4px;
+    float: left;
+    margin: 1em;
+    padding: 1em;
+    background-color: rgb(238, 238, 238);
+    box-sizing: border-box;
+    text-decoration: none;
+  }
+</style>
```
Comment (tomhughes) [inline app/views/events/show.html.erb]: Inline CSS is not allowed so this needs to be moved to a stylesheet.

File: db/structure.sql [hunk 5]
```diff
@@ -128,33 +114,6 @@ CREATE TYPE public.user_status_enum AS ENUM (
 );
 
 
---
--- Name: maptile_for_point(bigint, bigint, integer); Type: FUNCTION; Schema: public; Owner: -
---
-
-CREATE FUNCTION public.maptile_for_point(bigint, bigint, integer) RETURNS integer
-    LANGUAGE c STRICT
```
Comment (mmd-osm) [inline db/structure.sql]: why did you remove those functions?

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    </p>
+    <a href="https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=44.97067%2C-93.26822%3B38.90201%2C-77.04469#map=6/42.026/-85.156">Directions to this location.</a>
+    <%= image_tag("dc_library.png", size: "400x400") %>
+  </div>
+</div>
+
+
+
+
+<!--<%#= link_to 'Edit', edit_event_path(@event) %> |-->
```
Comment (tomhughes) [inline app/views/events/show.html.erb]: Don't just comment things out - if you don't need them then remove them.

File: app/views/microcosms/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,70 @@
+<%# if microcosm.errors.any? %>
+<!--    <div id="error_explanation">-->
+<!--      <h2><%#= pluralize(microcosm.errors.count, "error") %> prohibited this microcosm from being saved:</h2>-->
+<!--      <ul>-->
+          <%# microcosm.errors.full_messages.each do |message| %>
+<!--          <li><%#= message %></li>-->
+          <%# end %>
+<!--      </ul>-->
+<!--    </div>-->
+<%# end %>
```
Comment (tomhughes) [inline app/views/microcosms/_form.html.erb]: Commented out code again.

File: app/views/microcosms/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  </div>
+
+  <div class="field">
+    <%= form.label :lon %>
+    <%= form.text_field :lon, id: :microcosm_lon %> (decimal)
+  </div>
+
+  <p>
+    To get a bounding box use Geofabrik's <a href="http://tools.geofabrik.de/calc/#type=geofabrik_standard&bbox=-120.848658,31.674368,-79.375023,49.413863&tab=1&proj=EPSG:4326&places=2">Tile Calculator</a> tool.
+  </p>
```
Comment (tomhughes) [inline app/views/microcosms/_form.html.erb]: I don't know what the solution is offhand but I don't think we want to be referring people to a third party service like this.

File: app/views/microcosms/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    border: 1px solid #888;
+    border-radius: 4px;
+    float: left;
+    margin: 1em;
+    padding: 1em;
+    background-color: rgb(238, 238, 238);
+    box-sizing: border-box;
+    text-decoration: none;
+  }
+</style>
```
Comment (tomhughes) [inline app/views/microcosms/show.html.erb]: More inline styles.

File: db/migrate/20190901143302_create_friendly_id_slugs.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
+MIGRATION_CLASS =
+  if ActiveRecord::VERSION::MAJOR >= 5
+    ActiveRecord::Migration["#{ActiveRecord::VERSION::MAJOR}.#{ActiveRecord::VERSION::MINOR}"]
+  else
+    ActiveRecord::Migration
+  end
```
Comment (tomhughes) [inline db/migrate/20190901143302_create_friendly_id_slugs.rb]: Why on earth is this needed - we know what version of `ActiveRecord` we are starting from so there is no need for all this noise. I'm guessing it's automatically generated code, but we have no need for it.

File: db/structure.sql [hunk 11]
```diff
@@ -684,6 +643,107 @@ CREATE TABLE public.diary_entry_subscriptions (
 );
 
 
+--
+-- Name: event_attendances; Type: TABLE; Schema: public; Owner: -
+--
+
+CREATE TABLE public.event_attendances (
+    id bigint NOT NULL,
+    user_id integer,
```
Comment (mmd-osm) [inline db/structure.sql]: I'm somehow missing a FOREIGN KEY relationship in the code, e.g. `FOREIGN KEY (user_id) REFERENCES users(id)`

File: db/structure.sql (hunk 5 above)
Comment (tomhughes) [reply db/structure.sql]: That's just a result of not having them in his database when the structure was dumped - when running migrations you need to carefully review changes to the structure and remove any that aren't appropriate before committing.

File: db/structure.sql (hunk 11 above)
Comment (openbrian) [reply db/structure.sql]: This FK and any other missing ones will be added.

File: app/assets/javascripts/event_attendances.coffee (hunk 1 above)
Comment (openbrian) [reply app/assets/javascripts/event_attendances.coffee]: got it

File: app/helpers/event_attendances_helper.rb (hunk 2 above)
Comment (openbrian) [reply app/helpers/event_attendances_helper.rb]: ok

File: Design.md [hunk 12]
```diff
@@ -0,0 +1 @@
+/home/brian/Workshop/Projects/Microcosm/website/Design.md
```
Comment (mmd-osm) [inline Design.md]: Where is /home/brian?

File: app/views/events/show.html.erb (hunk 4 above)
Comment (openbrian) [reply app/views/events/show.html.erb]: Yes.  My first attack at groups (2018) used SASS (https://github.com/openbrian/microcosm-aoi/blob/master/app/assets/stylesheets/application.css.scss#L141 notice different repo here).  Is that ok?

File: app/views/microcosms/_form.html.erb (hunk 8 above)
Comment (openbrian) [reply app/views/microcosms/_form.html.erb]: Got it.

File: app/views/microcosms/show.html.erb (hunk 9 above)
Comment (openbrian) [reply app/views/microcosms/show.html.erb]: I was just telling Simon in matrix that this was done for the purpose of a SotM presentation.  I'll remove all inline style.

File: app/controllers/microcosms_controller.rb [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  def show_events
+  end
+
+  def edit
+  end
+
+  def update
+    respond_to do |format|
+      if @microcosm.update(microcosm_params)
+        format.html { redirect_to @microcosm, notice: 'Microcosm was successfully updated.' }
```
Comment (mmd-osm) [inline app/controllers/microcosms_controller.rb]: probably lots of those strings would be better off in locales for i18n.

File: app/views/events/show.html.erb (hunk 4 above)
Comment (tomhughes) [reply app/views/events/show.html.erb]: Yes SASS is fine and it can either go in there or you can create a new file in that directory.

It just can't be inline because our content security policy prohibits that. I'm surprised you didn't encounter that when testing to be honest.

**Comment by openbrian** [comment]
@tomhughes I had to make 1 change to a migration in order to get my database up and running.  I have not committed it.  20190518115041_add_acl_indexes.rb

-    add_index :acls, :address, :using => :gist, :opclass => :inet_ops
+    add_index :acls, :address, :using => :gist, :opclass => :gist_inet_ops

My PG version is 9.5.

File: config/settings/development.yml [hunk 14]
```diff
@@ -0,0 +1,2 @@
+osm_auth_id: "0ZlzyAW1vzK9HXQl9ekLUN07PIIO2uCrrpyODCYP"
```
Comment (mmd-osm) [inline config/settings/development.yml]: i think this file shouldn't be in this pull request

File: config/settings/production.yml [hunk 15]
```diff
@@ -0,0 +1,2 @@
+osm_auth_id: "0ZlzyAW1vzK9HXQl9ekLUN07PIIO2uCrrpyODCYP"
```
Comment (mmd-osm) [inline config/settings/production.yml]: same here.

File: db/migrate/20190903030453_add_location_to_microcosms.rb [hunk 16]
```diff
@@ -1,13 +1,16 @@
...
-    change_table "microcosms", :bulk => true do |t|
-      t.string "location", :null => false
-      t.decimal "lat", :null => false
-      t.decimal "lon", :null => false
-      t.integer "min_lat", :null => false
-      t.integer "max_lat", :null => false
-      t.integer "min_lon", :null => false
-      t.integer "max_lon", :null => false
+    # This group of migrations for microcosms will be run together.
+    safety_assured do
```
Comment (gravitystorm) [inline db/migrate/20190903030453_add_location_to_microcosms.rb]: `safety_assured` doesn't work like this! It's a warning that this migration could lock the table (potentially for hours, if there's lots of data) and so you need to find a different approach.

On a production database, this could be a multi-step process of creating shadow columns, backfilling data and so on. See https://github.com/ankane/strong_migrations for more details. However, since this is going to be an empty table, that would be overkill. 

So the best approach for pull requests is to avoid including multiple database migrations. Sure, during development you might realise that the initial database structure was incorrect, and add a fixup migration like this one. But for the PR, please make one migration with your preferred structure. For example, see #1576, where even though there were substantial changes to the database during the development of the PR, there was in the end only one migration.

File: app/controllers/event_attendances_controller.rb [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  before_action :authorize_web
+  before_action :set_event_attendance, :only => [:update]
+
+  authorize_resource
+
+  def create
+    attendance = EventAttendance.new(attendance_params)
+    attendance.intention = intention
+    if attendance.save!
+      redirect_to event_path(attendance.event), :notice => "Attendance was successfully saved."
```
Comment (gravitystorm) [inline app/controllers/event_attendances_controller.rb]: All these strings need to be through the i18n system. I find it best to do so right from the start, since it's time consuming to go back through all the controllers and views later on.

File: app/controllers/event_attendances_controller.rb [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+  layout "site"
+  before_action :authorize_web
+  before_action :set_event_attendance, :only => [:update]
+
+  authorize_resource
+
+  def create
+    attendance = EventAttendance.new(attendance_params)
+    attendance.intention = intention
+    if attendance.save!
```
Comment (gravitystorm) [inline app/controllers/event_attendances_controller.rb]: model.save! (with an exclamation mark) raises an error if the save is unsuccessful. So here you can use `attendance.save` (without the exclamation mark) which returns a boolean.

File: app/models/event_attendance.rb [hunk 19]
```diff
@@ -0,0 +1,21 @@
+# == Schema Information
+#
+# Table name: event_attendances
+#
+#  id         :bigint(8)        not null, primary key
+#  user_id    :integer          not null
+#  event_id   :integer          not null
+#  intention  :string           not null
```
Comment (gravitystorm) [inline app/models/event_attendance.rb]: I see "Yes" and "No" as the values for the intention. Are there likely to be other values? Would an enum be more useful here?

File: app/models/microcosm.rb [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+#
+#  id          :bigint(8)        not null, primary key
+#  name        :string           not null
+#  description :text
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#  slug        :string           not null
+#  location    :string           not null
+#  lat         :decimal(, )      not null
+#  lon         :decimal(, )      not null
```
Comment (gravitystorm) [inline app/models/microcosm.rb]: We normally use either integers (e.g. for nodes) or floats (e.g. for user home locations).

File: app/models/microcosm.rb [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+#  min_lat     :integer          not null
+#  max_lat     :integer          not null
+#  min_lon     :integer          not null
+#  max_lon     :integer          not null
+#
+
+class Microcosm < ActiveRecord::Base
+  extend FriendlyId
+  friendly_id :name, :use => :slugged
+  self.ignored_columns = ["key"]
```
Comment (gravitystorm) [inline app/models/microcosm.rb]: This is, I think, a side effect of having multiple migrations, so can be removed after the migrations are combined.

File: app/models/microcosm_member.rb [hunk 22]
```diff
@@ -0,0 +1,30 @@
+# == Schema Information
+#
+# Table name: microcosm_members
+#
+#  id           :bigint(8)        not null, primary key
+#  microcosm_id :integer          not null
+#  user_id      :integer          not null
+#  role         :string(64)       not null
```
Comment (gravitystorm) [inline app/models/microcosm_member.rb]: Could be an enum

File: app/views/events/show.html.erb [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+      Organized by: <strong>Andrew Wiseman</strong>
+    </p>
+  </div>
+  <div>
+    <p><%= @event.attendees.size %> people are going.</p>
+    <p>Are you going?</p>
+    <%= form_with :model => @my_attendance do |form| %>
+      <%= form.hidden_field(:event_id, :value => @event.id) %>
+      <%= form.hidden_field(:user_id, :value => current_user&.id) %>
+      <%= form.submit :value => "Yes" %>
```
Comment (gravitystorm) [inline app/views/events/show.html.erb]: We'll need to change this so that what the user sees is via i18n, while the value submitted doesn't change

File: app/views/events/show.html.erb [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+    </div>
+
+    <h1 class="title">
+      <%= @event.title %>
+    </h1>
+    <p>
+      Hosted by <strong><%= link_to @event.microcosm.name, microcosm_path(@event.microcosm) %></strong>
+    </p>
+    <p>
+      Organized by: <strong>Andrew Wiseman</strong>
```
Comment (gravitystorm) [inline app/views/events/show.html.erb]: Placeholder text needs updating

File: test/factories/events.rb [hunk 25]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :event do
+    title { "MyString" }
```
Comment (gravitystorm) [inline test/factories/events.rb]: Why is everything hardcoded in the factories?

File: app/views/microcosms/_form.html.erb (hunk 8 above)
Comment (Nakaner) [reply app/views/microcosms/_form.html.erb]: You might use https://github.com/kajic/leaflet-locationfilter to allow users select the bounding box. The Leaflet plugin is used at https://tyrasd.github.io/osm-qa-feeds/ (you have to search for a location first).

Comment (tomhughes) [reply app/views/microcosms/_form.html.erb]: We already have leaflet-location filter installed - it is used for selecting the area to export.

**Comment by gravitystorm** [comment]
> I mean sure if we were starting now we might move to cucumber

I'm going to go further, and say that I don't want to move to cucumber at all, even for new stuff, or even as a long-term goal. I find the cucumber tests hard to read, much more so than normal ruby-code integration/system tests.  I don't find it useful to write "When I visit the home page" instead of `visit root_path`. And I don't think we have any non-technical people reading the tests anyway.

https://www.jackkinsella.ie/articles/why-bother-with-cucumber-testing is a relevant read.

So I'd prefer to see the cucumber tests implemented as normal system tests.

**Comment by gravitystorm** [comment]
> I'd also point out that we generally advise opening discussions on large scale features before doing lots of work on them so that basic structural issues, naming, etc can be resolved before lots of code is written.

Yeah, I'd agree with this. I find it very hard to figure out the basics here because I get distracted by all the little things (like my notes in my review today, or the tests, or whatnot) and I can't reason about the bigger picture. My approach to the Moderation PR was the same - I had to fix all the little things before the bigger picture became clear to me. I didn't mind that when it was my own time spent on the fixups, but I feel bad about asking other people to do fixups on code that might end up disappearing if we change the bigger picture.

So it might be worth going back to basics on this, and discussing the data model, whether it's something we want, whether some bits are optional and can be moved to followup PRs, etc.

> I mean this is definitely something that people want, but whether `Microcosm` is the right name is something we need to discuss but would now involve a large amount of work.

I think the naming comes from it being initially developed as a standalone project, rather than being developed from the start as a feature of this site. So it's great to see that @openbrian is working on integrating it!

**Comment by mmd-osm** [comment]
In case you missed it - this might be interesting to get the bigger picture and some functional level discussions. 

Brian DeRocher 19:12
For those interested there's some discussion of OSM groups on matrix (https://riot.im/app/#/room/#osmcommunity:matrix.org) (edited) 

Is this code up for testing somewhere?

**Comment by openbrian** [comment]
> Is this code up for testing somewhere?

@mmd-osm Not yet.  But it will be soon.  I need to rip out hard coded things for my SotM US presentation.

I created a separate branch to add oauth via osm.org, so people's OSM account will get them into this demo site.  I'll create a "deployment" branch and merge these branches together before deploying to my server.

I want to use Docker, but not sure that work is ready to use yet.

Let me know if you all are interested in that omniauth branch.  I can push in a separate PR.  It's much like the previous one added for wikipedia.

**Comment by mmd-osm** [comment]
@tomhughes : I wonder if https://github.com/openbrian/osm-microcosms could be safely deployed on the dev instance already, and if not, what parts still wound need some work. 

@openbrian : I wouldn't spend too much time on pulling in users from osm.org via oauth, at least  in the current development phase. Creating some new test account on the (or some) dev instance looks perfectly fine to me.

File: app/views/microcosms/_form.html.erb (hunk 8 above)
Comment (openbrian) [reply app/views/microcosms/_form.html.erb]: For now we can just just grab numbers from "Export" button at top of osm.org.

File: app/controllers/event_attendances_controller.rb (hunk 17 above)
Comment (openbrian) [reply app/controllers/event_attendances_controller.rb]: yup yup

File: app/controllers/event_attendances_controller.rb (hunk 18 above)
Comment (openbrian) [reply app/controllers/event_attendances_controller.rb]: Fixed this type of issue in 3 places.

File: app/models/event_attendance.rb (hunk 19 above)
Comment (openbrian) [reply app/models/event_attendance.rb]: Likely there will be at least 1 more intention, something like "Interested".  But finitely many and an enum would still make sense.

File: db/migrate/20190903030453_add_location_to_microcosms.rb (hunk 16 above)
Comment (openbrian) [reply db/migrate/20190903030453_add_location_to_microcosms.rb]: I'll collect migrations later.  I created a ticket to remind me.  https://github.com/openbrian/osm-microcosms/issues/3

Comment (openbrian) [reply db/migrate/20190903030453_add_location_to_microcosms.rb]: Rake bitched at me because I was adding columns and setting the default to null at the same time.  Of course that's not going to work on a table that's not empty.  I commented that safety was assured because this migration and the one to create the table would be run together.  When collapsing migrations this goes away.

File: app/models/microcosm.rb (hunk 21 above)
Comment (openbrian) [reply app/models/microcosm.rb]: yes

File: app/models/microcosm.rb (hunk 20 above)
Comment (openbrian) [reply app/models/microcosm.rb]: Yeah, you can see below in this file the mins and maxes are integers.  Not sure how these two went rogue.  Fixed.

**Comment by tomhughes** [comment]
Please don't pollute the commit history with nonsense - if the previous build failed then adding some whitespace to trigger a new one isn't going to fix anything.

If you need help understanding why the build is failing then please ask instead of randomly moving lines around and inserting whitespace.

Likewise if you really need a new build triggering then ask as I can trigger a rebuild without polluting the commit history.

**Comment by openbrian** [comment]
I understand the builds are isolated and deterministic.  I just activated Travis builds for my fork (https://travis-ci.org/openbrian/osm-microcosms/).

**Comment by mmd-osm** [comment]
As I was setting up a local Docker container, I thought we could add a few screenshots to highlight how microcosms look like at the moment. So here we go:

## List of all microcosms

all_microcosms

## Details for one microcosm

single_microcosm

## Creating a new microcosm


new_microcosm

## Microcosm member list

microcosm_members

## Edit microcosm member list

edit_membership

## Creating a new microcosm event

creating_new_event
creating_new_event2


## Joining a microcosm event

join_event

**Comment by openbrian** [comment]
Beautiful, aside from all the hard coded thing :roll_eyes: I've been busy lately with upgrading my server, but when I get back into OSM, I'll be undoing the hard coded things and resolving more issues from above.  Thanks @mmd-osm .

**Comment by gravitystorm** [comment]
@openbrian just checking in here to see what your plans are for this PR, and microcosms in general?

**Comment by openbrian** [comment]
My plan is to continue to address issues found with it, add more features, and repeat.  I'm a little tied up with another project, but when I get more free time, I look forward to working on this again.

**Comment by RobJN** [comment]
@openbrian what can be done to help to get this finished? I ask because many of the Local Chapters & Communities Working Group (LCCWG) members were and still are fans of this. We see it as a helpful way for mappers to build communities without having to leave osm.org. We skipped our December meeting but can pick this up again in January - it would be great if we had some options to discuss.

File: app/views/events/show.html.erb (hunk 24 above)
Comment (openbrian) [reply app/views/events/show.html.erb]: Fixed

**Comment by gravitystorm** [comment]
Great to see more progress on this @openbrian !

I saw some things that caught my eye in the last few updates, for example:

* hardcoding webdriver stuff in the travis config, vs e.g. https://github.com/titusfortner/webdrivers 
* even more cucumber tests being added, vs using our existing test suites
* lots of CSS and hardcoded layout stuff, vs using bootstrap for responsive layouts
* other stuff we've discussed previously, like hardcoded values in factories

I want to try to nudge you towards making this PR something that we can discuss and merge, rather than it ever-expanding project that is seemingly not approaching a merge-able state! Basically I have worries that this is following the same path as the GSoC Moderation branch, where it took me 11 months of work on the PR after the original contributor had moved on, and I don't want the same thing to happen again here.

What do you think? What can I do to help?

**Comment by openbrian** [comment]
@gravitystorm Thanks for the continued feedback and support.

- I don't like the hard coded webdriver either.  I see https://github.com/openstreetmap/openstreetmap-website/blob/master/INSTALL.md suggests to install phantomjs.  But to my understanding is that phantomjs has been retired.  So I installed geckodriver instead.  I'll see how titusfortner/webdriver works out.
- I'm going to continue using cucumber as that's how I get from the design document Design.md to test scenarios, to specs, to code.  I'm not opposed to minitest.  In fact I'm a strong supporter of the testing pyramid, which places unit tests as the most popular testing in your suite.  i just need to know where to get started.
- Bootstrap, got it.
- Hard coded values in factories comes from rails-generated code.  There are no minitests yet, so when I start writing them, I'll remove hard coded values.

Do you think we haven't been discussing this PR?  I've been methodically going through comments above and addressing each one of them.   I haven't added a single new feature, aside from replacing hard coded things (e.g. the map image) with real things (leaflet map).

I'm sorry you were burned by a previous experience with a developer that left.  But I've been involved with OSM since 2009 and running MappingDC for nearly a decade.  I've also been on the board of OSM US.  So I hope you feel I'm not just going to up and leave.  After all I'm building this feature so **I** can use it too.  I'm the one shelling out $200 per year to Meetup.

You can do 3 things to help.
1. Get me started with minitest.  Feel free to fork my fork and submit a PR.
2. Set me up with a dev server to we can do some actual user centered design.
3. Many of my pages will have styles specific to that page.  Where do you come down on page specific files versus common.scss?  I personally feel common.scss is bloated and unorganized.  It doesn't seem to me that .edit_trace, #userinformation, .activity-details, .diary_post, or .users-terms are common as most of them are used only once.  (Side note, I don't see users-terms being used anywhere.)

**Comment by gravitystorm** [comment]
> So I hope you feel I'm not just going to up and leave.

That's not my main concern - my main worry is that if this PR continues to diverge from upstream, neither of us will find it particularly enjoyable to spend months refactoring it! For example, every time you write a cucumber test, I see something that will need to be rewritten before it can be reviewed. Same with the other topics that I mention. So it's less work all round if we sort things like that out sooner rather than later. That's what I mean by heading towards a merge-able state.

I'll talk more about cucumber specifically in a separate issue, so that we don't swamp this one.

I can open a ticket to ask for your branch to be deployed on apis.dev.osm.org

With styles, we now have Bootstrap available and the need for custom styles for any particular page will be greatly reduced. I would aim for you having no custom CSS at all, and instead picking components and utilities from the list available at https://getbootstrap.com/docs/4.4/getting-started/introduction/ and otherwise letting bootstrap do the styling by default. The current common.css file is a mess and most of it is disappearing as I continue refactoring the site, so it's best ignored.

File: app/models/event_attendance.rb (hunk 19 above)
Comment (openbrian) [reply app/models/event_attendance.rb]: Created https://github.com/openbrian/osm-microcosms/issues/8 to track this.

File: app/models/microcosm_member.rb (hunk 22 above)
Comment (openbrian) [reply app/models/microcosm_member.rb]: Created https://github.com/openbrian/osm-microcosms/issues/9 to track this.

**Comment by openbrian** [comment]
@gravitystorm I'm following up on the change to replace geckodriver in .travis.yml with using the titusfortner/webdrivers gem.  Somehow it works locally (yes, I've taken geckodriver out my $PATH).  But tests fail on travis because webmock is preventing the download of the webdrivers.

I looked back in the git history to see where webmock was introduced and why.  Seems to be used in test_helper.rb in with_http_stubs().  So instead of calling out to geonames, geocoder_ca, and nominatim, we just return data from disk.

The policy to disable network connection (except for localhost) is also in test_helper.rb line 24.

Looks like webmock allows a whitelist of allowed sites.  https://www.rubydoc.info/gems/webmock/2.0.2#External_requests_can_be_disabled_while_allowing_specific_requests  I would add for example https://chromedriver.storage.googleapis.com

Do you suggest doing that?

**Comment by tomhughes** [comment]
Calling out to external sites in tests is a terrible idea - it means they can randomly break because an external site is down.

**Comment by gravitystorm** [comment]
> Do you suggest doing that?

I'd suggest spinning this part out into a separate PR, so that we can discuss it and refine it there.

File: Design.md [hunk 26]
```diff
@@ -53,6 +53,7 @@ There are various hosting options:
 - [x] See upcoming events for a microcosm
 - [x] See all upcoming events
 - [ ] See past events
+- [ ] Get directions to teh event
```
Comment (HolgerJeromin) [inline Design.md]: "The" typo

File: Design.md [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+* https://github.com/keepright/keepright
+* http://osmose.openstreetmap.fr/en/map/
+* https://wiki.openstreetmap.org/wiki/OSM_Inspector
+
+
+# Use Cases
+
+## List reviews in the area
+
+Some people want their changes reviewed.  Provide a list of these for the AOI.
```
Comment (HolgerJeromin) [inline Design.md]: "API" typo?

File: config/locales/en.yml [hunk 28]
```diff
@@ -6,19 +6,51 @@ en:
...
       blog: "%e %B %Y"
   helpers:
+    label:
+      event:
+        description: "Description"
+        location: "Location"
+        location_url: "Location URL"
+        microcosm_id: "Microcosm"
+        moment: "When"
+        submit: "Create Event2"
```
Comment (HolgerJeromin) [inline config/locales/en.yml]: Is the "2" correct?

File: Design.md (hunk 27 above)
Comment (openbrian) [reply Design.md]: Area of interest - the area that a microcosm covers.

File: Design.md (hunk 26 above)
Comment (openbrian) [reply Design.md]: Fixed.  Thank you.

File: config/locales/en.yml (hunk 28 above)
Comment (openbrian) [reply config/locales/en.yml]: The whole submit is not correct.  The submit field/button doesn't have a <label>.  I have to look into internationalizing the text on the submit button as it behaves differently from other form fields.  Thanks.

**Comment by gravitystorm** [comment]
>     * I'm going to continue using cucumber as that's how I get from the design document Design.md to test scenarios, to specs, to code.  I'm not opposed to minitest.  In fact I'm a strong supporter of the testing pyramid, which places unit tests as the most popular testing in your suite.  i just need to know where to get started.

>     1. Get me started with minitest.  Feel free to fork my fork and submit a PR.

I'm afraid that I don't have the spare capacity right now to refactor your PR, but I can certainly point you in the right direction! For any of your tests that involve user interaction - the sort of thing where you've been using cucumber and capybara - then you should refactor the tests to be "system" tests in the `test/system' directory. You can see some examples there, and continue to use all the capybara stuff like `visit login_path`, `click_on title` and so on in those tests.

**Comment by openbrian** [comment]
I just merged from upstream.  All tests passing.

**Comment by gravitystorm** [comment]
@openbrian Let me know when you would like a review. I can see a few things that we should discuss, but I think you're still iterating on things. My preference would be to review soon and merge incrementally.

**Comment by openbrian** [comment]
I'm going to write more tests and cut back on a couple features.

**Comment by openbrian** [comment]
Dev site is down.  Any chance you can get it up asap.

**Comment by tomhughes** [comment]
Rebuilding the assets and restarting the application seems to have fixed it.

File: app/controllers/microcosms_controller.rb (hunk 13 above)
Comment (openbrian) [reply app/controllers/microcosms_controller.rb]: Got the translation in there.

File: test/factories/events.rb (hunk 25 above)
Comment (openbrian) [reply test/factories/events.rb]: They are now all generating unique/random/custom data.

**Comment by tordans** [comment]
For all that are following this project, there is now a recording of Brians SOTMUS Talk online at https://www.youtube.com/watch?v=ZcTgz6pZ65U. – Thanks for that Brian!

And there is an instance running at https://mappingdc.org/microcosms to try out and even use (but right now I get a `502 Bad Gateway`).

---

@openbrian in you talk it sounded like you are now pushing for a first release and also want to focus on integrating the work in osm.org (both is great IMO). My take from what I saw in your talk is, that there are quite a few topics (small and big) that need to be figured out first. Thinks like styling; navigation integration; the calendar feature vs. osmcal.org; possible links to external sites; possibly performance testing; those community pages vs. the WIP community page #3301 – to name a few that came to mind.

What would be a good way to start the conversation about resolving those topics? Maybe it's time for a zoom call with @gravitystorm to create a list of things that we need to resolve to get to a mergable state?

And – just to put it out there – I think it worth considering/discussing splitting this project off of osm.org. As far as I see, it could be a standalone rails app that uses the Oauth API (maybe the same way that the new forums does) to authenticate users. However, that might require to extend a few API endpoints to support the filtering you do on your microcosm-profile-pages.

**Comment by mmd-osm** [comment]
@openbrian : Is this PR still needed, or has it been superseded by subsequent PRs?

**Comment by gravitystorm** [comment]
Superseded by https://github.com/openstreetmap/openstreetmap-website/pull/3717
